google/cloud/bigtable/internal/async_row_sampler_test.cc

```diff
@@ -0,0 +1,449 @@
+              }
+            })
+            .WillOnce([](btproto::SampleRowKeysResponse*, void*) {});
+        EXPECT_CALL(*reader, Finish).WillOnce([](grpc::Status* status, void*) {
+          *status = grpc::Status::OK;
+        });
+        return reader;
+      });
+
+  auto samples_future = table_.AsyncSampleRows();
+
+  // Start()
+  cq_impl_->SimulateCompletion(true);
+  // Return response
+  cq_impl_->SimulateCompletion(true);
+  // End stream
+  cq_impl_->SimulateCompletion(false);
+  // Finish()
+  cq_impl_->SimulateCompletion(true);
+  // Simulate the backoff timer
+  cq_impl_->SimulateCompletion(true);
+
+  ASSERT_EQ(1U, cq_impl_->size());
+
+  // Start()
+  cq_impl_->SimulateCompletion(true);
+  // Return response
+  cq_impl_->SimulateCompletion(true);
+  // End stream
+  cq_impl_->SimulateCompletion(false);
+  // Finish()
+  cq_impl_->SimulateCompletion(true);
+
+  ASSERT_EQ(0U, cq_impl_->size());
+
+  auto status = samples_future.get();
+  ASSERT_STATUS_OK(status);
+
+  auto samples = status.value();
+  ASSERT_EQ(1U, samples.size());
+  EXPECT_EQ("test2", samples[0].row_key);
+  EXPECT_EQ(22, samples[0].offset_bytes);
+}
+
+TEST_F(AsyncSampleRowKeysTest, TooManyFailures) {
+  // We give up on the 3rd error.
+  auto constexpr kErrorCount = 2;
+  Table custom_table(client_, "foo_table",
+                     LimitedErrorCountRetryPolicy(kErrorCount));
+
+  EXPECT_CALL(*client_, PrepareAsyncSampleRowKeys)
+      .Times(kErrorCount + 1)
+      .WillRepeatedly([](grpc::ClientContext*,
+                         btproto::SampleRowKeysRequest const&,
+                         grpc::CompletionQueue*) {
+        auto reader = absl::make_unique<
+            MockClientAsyncReaderInterface<btproto::SampleRowKeysResponse>>();
+        EXPECT_CALL(*reader, StartCall);
+        EXPECT_CALL(*reader, Read).Times(2);
+        EXPECT_CALL(*reader, Finish).WillOnce([](grpc::Status* status, void*) {
+          *status = grpc::Status(grpc::StatusCode::UNAVAILABLE, "try again");
+        });
+        return reader;
+      });
+
+  auto samples_future = custom_table.AsyncSampleRows();
+
+  for (int retry = 0; retry < kErrorCount; ++retry) {
+    // Start()
+    cq_impl_->SimulateCompletion(true);
+    // Return response
+    cq_impl_->SimulateCompletion(true);
+    // End stream
+    cq_impl_->SimulateCompletion(false);
+    // Finish()
+    cq_impl_->SimulateCompletion(true);
+    // Simulate the backoff timer
+    cq_impl_->SimulateCompletion(true);
+
+    ASSERT_EQ(1U, cq_impl_->size());
+  }
+
+  // Start()
+  cq_impl_->SimulateCompletion(true);
+  // Return response
+  cq_impl_->SimulateCompletion(true);
+  // End stream
+  cq_impl_->SimulateCompletion(false);
+  // Finish()
+  cq_impl_->SimulateCompletion(true);
+
+  auto status = samples_future.get();
+  ASSERT_THAT(status,
+              StatusIs(StatusCode::kUnavailable, HasSubstr("try again")));
+
+  ASSERT_EQ(0U, cq_impl_->size());
+}
+
+class MockBackoffPolicy : public RPCBackoffPolicy {
```

Comment:
Consider refactoring this, it now appears in several tests.... Also move to the top.  If that is too much work for this PR, create an issue to clean this up.